atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -15,19 +15,361 @@
+        RangeRequest range = RangeRequest.builder()
+                .startRowInclusive(startRow)
+                .batchHint(rowBatchSize)
+                .build();
+
+        try (ClosableIterator<RowResult<Value>> valueResults = getValues(tableRef, range, sweepTs, sweeper);
+             ClosableIterator<RowResult<Set<Long>>> rowResults = keyValueService.getRangeOfTimestamps(
+                     tableRef, range, sweepTs)) {
+            EquivalenceCountingIterator<RowResult<Set<Long>>> rowResultTimestamps =
+                    new EquivalenceCountingIterator<>(rowResults, rowBatchSize, sameRowEquivalence());
+            PeekingIterator<RowResult<Value>> peekingValues = Iterators.peekingIterator(valueResults);
+
+            BatchingVisitable<CellAndTimestamps> cellsAndTimestamps = BatchingVisitableFromIterable
+                    .create(getTimestampsFromRowResultsIterator(() -> rowResultTimestamps));
+
+            final AtomicInteger totalCellsSwept = new AtomicInteger(0);
+            cellsAndTimestamps.batchAccept(
+                    cellBatchSize,
+                    thisBatch -> {
+                        CellsAndTimestamps thisBatchCells = CellsAndTimestamps.fromCellAndTimestampsList(thisBatch);
+                        int cellsSwept = sweepForCells(thisBatchCells,
+                                tableRef,
+                                sweeper,
+                                sweepTs,
+                                peekingValues,
+                                runType);
+                        totalCellsSwept.addAndGet(cellsSwept);
+                        return true;
+                    });
+
+            byte[] nextRow = rowResultTimestamps.size() < rowBatchSize ? null :
+                    RangeRequests.getNextStartRow(false, rowResultTimestamps.lastItem().getRowName());
+            return SweepResults.builder()
+                    .previousStartRow(Optional.fromNullable(startRow))
+                    .nextStartRow(Optional.fromNullable(nextRow))
+                    .cellsExamined(rowResultTimestamps.size())
+                    .cellsDeleted(totalCellsSwept.get())
+                    .sweptTimestamp(sweepTs)
+                    .build();
+        }
+    }
+
+    private ClosableIterator<RowResult<Value>> getValues(TableReference tableRef,
+                                                         RangeRequest range,
+                                                         long sweepTs,
+                                                         Sweeper sweeper) {
+        if (sweeper.shouldSweepLastCommitted()) {
+            return keyValueService.getRange(tableRef, range, sweepTs);
+        } else {
+            return ClosableIterators.emptyImmutableClosableIterator();
+        }
+    }
+
+    public static Equivalence<RowResult<Set<Long>>> sameRowEquivalence() {
+        return new Equivalence<RowResult<Set<Long>>>() {
```

Comment:
Again, I didn't touch this. Also, this will be gone after my refactor (stay tuned).